### ISSUE TITLE
Add `register_action` decorator

### ISSUE THREAD

**Issue by lucyleeow** [Napari/napari#7312]
This is mentioned in #7056 but warrants it's own issue.

With app-model migration, NAP-7 and the over haul of the keybinding implementation we are thinking of adding a `register_action` decorator, and deprecate [`bind_key`](https://github.com/napari/napari/blob/a4fa1fa00ecee4a492daf77d7f901a4e4acc2734/napari/utils/key_bindings.py#L134). Reasons

* Registering the action with app-model allows the command to be tracked by app model, which will be useful for:
   * keeping track of keybindings, resolving conflicts and potentially showing them in the shortcuts tab (? not sure on this)
   * user can also specify keybinding enablement (i.e. `when`/context)
   * enable users to use the common API (#7190) for executing commands, which is also useful for future 'redo/undo/workflow' features
* Users could add the action to a menu at the same time
* Command enablement expression could also be specified 

Concerns:

* Not declarative

This issue is intended to document everything. We will have to try and get comments from people on zulip/at community meetings on top.

cc @DragaDoncila @psobolewskiPhD

**Comment by lucyleeow** [comment]
The intention is for users to use this programmatically, and not for plugin users to use this, so you are right that this 'can't' be declarative.
@DragaDoncila mentioned this con, not sure if there was more to it?

**Comment by DragaDoncila** [comment]
There is indeed more to it.

The current discussions about keybindings only considers full commands and keybindings declared in the manifest. I believe that many plugin developers are using the current decorator within more complex widgets, binding keybindings to various sub-widgets and sub-actions. 

Such usage of keybindings will never make it to the manifest, because you can't bind the command until the widget object is created etc. In theory, this would be the gap that a `register_action` decorator would fill. I think this use-case is extremely common, and probably far more common than attaching keybindings to commands/full functions.

> We may implement this in this way that it will work only in scripts, in plugins it may print a warning ad do nothing.

I therefore think we **do** need to implement this decorator for use within widgets etc., but still in a way that allows us to capture it for the shortcuts tab.

**Comment by Czaki** [comment]
Clarification after meeting.

I think that plugins should be allowed only for declarative shortcuts. 
However declarative shortcuts are not possible in scripts or notebooks (or napari console). 

So we may provide such decorator, but only working outside plugins. 
For plugins, we may disable it (for example using context manager) and print a warning with instruction how to declare shortcuts in manifest

**Comment by jni** [comment]
> Such usage of keybindings will never make it to the manifest, because you can't bind the command until the widget object is created etc.

I don't think this is strictly necessarily true. Since *we* are the ones creating instances from plugins, plugins can declare keybindings that map to a class method, and that activate only when an instance has been created. We would then find that method within that specific instance and call it.

**Comment by jni** [comment]
(And, implied in my comment: I think it should be a design goal that everything a plugin provides is declarative. Maybe sometimes we have to fall short, but let's work towards it.)

**Comment by lucyleeow** [comment]
Notes from quick chat with @DragaDoncila about declaring keybinding to widget, sub-widget:

* Enablement - we need to have a context key so the keybinding is only available when the widget is available. I think this would be when the widget is visible (note that you can 'hide' and 'close' widgets, so I think we only want this keybinding when the widget is visible)
* De-register - we will need to de-register the action when the widget is closed. Potentially there could be an event relating to change/closing of widgets

**Comment by psobolewskiPhD** [comment]
`bind_key` is really intuitive for what it does and has been around a while. Any way we can change the backend but keep that as the API?

> Enablement - we need to have a context key so the keybinding is only available when the widget is available. I think this would be when the widget is visible (note that you can 'hide' and 'close' widgets, so I think we only want this keybinding when the widget is visible)

This would be great I think, it could mitigate some of the conflict issues! 
E.g. using an annotation plugin, it could "hijack" 1, 2, 3, etc. from the Layer keybindings for its own purposes when active. Then when closed, the normal keybinds would resume working.

Not sure how to do a nice job in the GUI -- VS Code has things like this, but it's basically "have fun editing some json" for a UI.

**Comment by lucyleeow** [comment]
> bind_key is really intuitive for what it does and has been around a while.

Thanks for the input, its good to know how often things are used 'in the wild'.
The intention with changing it to `register_action` is for the user to be able to add extra things to the action beyond just a keybinding:

* enablement for the function (currently would just be viewer/layer context `when`'s)
* ability to add the action to a menu (do you think this would be interesting? I thought of this just because app-model easily offers this)

WDYT?

> E.g. using an annotation plugin, it could "hijack" 1, 2, 3, etc. from the Layer keybindings for its own purposes when active. Then when closed, the normal keybinds would resume working.

I like this, and I think it makes sense but I feel like we should see what others think. Conflict handling could get tricky though, because 2 plugins could want to use the same keybinding for their widget BUT they won't necessarily conflict as the 2 wigets aren't necessarily going to be open at the same time. I guess if they are open at the same time we could do the pop-up. But then these are not strict conflicts, users could probably just specify priority. It gets complicated...

Just a note, for me, I even think it makes sense for layer keybindings to take precedence over viewer keybindings (e.g., when an image is active keybinding 'A' does something to the image and when no image layer is active we default back to a 'viewer' keybinding) but I think we didn't like this (see: #7210) .

**Comment by psobolewskiPhD** [comment]
> The intention with changing it to `register_action` is for the user to be able to add extra things to the action beyond just a keybinding:

Yeah, that totally makes sense, but I suspect a lot of people just want to bind a key to a function, much like they do now. So this extra features is a form of extra complexity that may not always be needed. To be clear, I'm not saying don't implement this, just thinking if we can keep `bind_key` using the new system for the simple case.

> Just a note, for me, I even think it makes sense for layer keybindings to take precedence over viewer keybindings (e.g., when an image is active keybinding 'A' does something to the image and when no image layer is active we default back to a 'viewer' keybinding) but I think we didn't like this (see: #7210) .

There's precedent for this isn't there: we have layer keybinds that do different things on different layers -- so they are contextually aware. I'd be ok with extending that to the viewer, but there are viewer things you want to keep working with any/all layers, e.g. reset view. It would be wierd if that stopped working because you had a Labels layer open or something? I guess layer <-> viewer conflicts we could have the warning include: "It's fine, I want it this way".

**Comment by lucyleeow** [comment]
>  I suspect a lot of people just want to bind a key to a function, 

Thanks, exactly the kind of info I'm after. I'm okay with supporting both `bind_key` and `register_action` but I do think it's less ideal to have 2 ways of doing things that differ only slightly. We could also only have bind key and don't offer for users to add to menu, as you say it may not even be of interest to users.

> I guess layer <-> viewer conflicts we could have the warning include: "It's fine, I want it this way".

Definitely an option. I think we can keep things as they are and if there is need from the community we can add the extra complexity.

**Comment by lucyleeow** [comment]
> To be clear, I'm not saying don't implement this, just thinking if we can keep bind_key using the new system for the simple case.

Just thought to clarify; we would do a deprecation cycle with `bind_key`, supporting both `bind_key` and `register_action` for some number of releases and giving a deprecation warning when `bind_key` is used, asking the user to switch to `register_action`

I agree that `bind_key` is more intuitive, in that it is pretty clear what you are doing.

With the switch to app-model though, the `bind_key` term does not convey well the extra notion that the action will be registered with app-model, you'll be able to execute it via app-model (https://github.com/napari/napari/issues/7190), which also allows us to track command history (useful for future redo/undo/workflow features) etc.

**Comment by DragaDoncila** [comment]
Coming back to this with some thoughts!

>  we need to have a context key so the keybinding is only available when the widget is available.

agreed, and I we need to think about what happens when a plugin declares a keybinding for a command that is associated with a widget contribution and **doesn't add this context key**.  Should we just always add this context key to the registered action for the developer? Should we invalidate manifests that don't declare this context key? I'm leaning towards the first option, except that maybe the second option is nice for education purposes and to make the developer explicitly consider what's going on with this keybinding? Or perhaps we add it for them **and** warn? IDK

> we will need to de-register the action when the widget is closed. Potentially there could be an event relating to change/closing of widgets

afaik all qt widgets would emit some sort of close event that we can connect to, so it'll be up to us to make sure we handle that event and de-register the corresponding action, possibly in the dock widget's own `closeEvent` handler? Only thing I'm not sure about is how to get the action's ID in there, but I guess we could store it as an attribute on the dock widget we build idk.

> it could mitigate some of the conflict issues! ... Not sure how to do a nice job in the GUI -- VS Code has things like this, but it's basically "have fun editing some json" for a UI

Yeah the UI is going to be a nightmare. VS Code does have a very nice preferences editor, but it's worth noting that as far as Lucy and I can tell, **it doesn't really do conflict resolution**. It prioritises extension keybindings at the same level as its own, allows you to set a weight for the keybinding, and then when it comes time to handle a keypress, it just lifts the first registered command that satisfies the context out of its already prioritized queue. In the preferences window, you can set conflicting keybindings to your heart's content. The only utility for managing them is you can right-click on a keybinding and then show all commands with the same keybinding. It's up to you to decide which you keep, and whether their contexts conflict.

For our purposes I think deciding on conflicts when there's no context keys and keeping the current layer and viewer conflicts is ok, but I would be very hesitant to add conflict checking for any more complex contexts than that e.g. different widgets being visible... I would be happy to warn when the user triggers such a keybinding (concrete example, two widgets are open that both declared the same keybinding) and say "hey multiple commands are active for this keybinding and we're running this one. Go to preferences to change".

> To be clear, I'm not saying don't implement this, just thinking if we can keep bind_key using the new system for the simple case.

It doesn't seem like that big a deal to me to have the `register_action` decorator and then also keep `bind_key` and have it call `register_action` with relevant defaults under the hood. The only thing we'd have to be clear about is any use cases that are no longer supported I guess. Do we have a concrete list of what those are...?

**Comment by lucyleeow** [comment]
> Should we just always add this context key to the registered action for the developer? Should we invalidate manifests that don't declare this context key? I'm leaning towards the first option, except that maybe the second option is nice for education purposes and to make the developer explicitly consider what's going on with this keybinding? Or perhaps we add it for them and warn? IDK

I would vote for having it automatically added. It reduces burden when dev is writing the plugin and chance of errors when writing the expression. It does also allow us to change how we implement contexts (re #7101 :sweat_smile: ) and thus what the expression would look like. Though now that I think about it, I am assuming we would check in the dockwidgets for a widget of that name, not sure what the 'context expression' would be in this case if we just use a function? 

With regard to:

> to make the developer explicitly consider what's going on with this keybinding

We can ensure that we document well (add comments to cookie-cuttter, update our plugins docs). I am assuming you are just talking about the fact that the action is only active when the widget is visible?

>  and warn

warn the user or the plugin dev? At what point?

> I'm not sure about is how to get the action's ID in there, but I guess we could store it as an attribute on the dock widget we build idk.

My understanding is that you can close the 'inner' widget and the dock widget, and in both cases you should de-register the actions right?

If you are closing the dock widget, you should just de-register all widget associated functions? Maybe we can just ensure these all have some prefix?

And if you are closing the inner widget, can you get the name of the widget?

> It doesn't seem like that big a deal to me to have the register_action decorator and then also keep bind_key and have it call register_action with relevant defaults under the hood. 

My concern here is that it makes things complicated. E.g., when you teach, would you  recommend the `bind_key` decorator but give the caveat that if they want to do more complex things, use `register_action`? And you would want to clear up potential confusion and make sure users know that both do the same thing, and both will register the action with app-model, `register_action` just has more functionality. Even with good documentation, there may still be confusion.

But yes your suggestion here is nice in that there would be limited extra maintenance burden. Again I am fine either way.

**Comment by DragaDoncila** [comment]
> I would vote for having it automatically added. It reduces burden when dev is writing the plugin and chance of errors when writing the expression.

Yeah that's fine, I think that makes sense. We'd just have to be explicit in the docs that "keybindings added to commands that are methods of widgets will only be active when the widget is visible". 

> not sure what the 'context expression' would be in this case if we just use a context function

I think you're right that it would be a function, since otherwise we'd need a context key for each named widget which is never going to work.

> My understanding is that you can close the 'inner' widget and the dock widget, and in both cases you should de-register the actions right?

I'm not sure what you mean by "dock widget" and "inner widget" here. All "dock widgets" have "inner widgets" consisting of the actual widget the plugin developer created, but I don't think you can close/hide **just** the "inner widget". Plugin developers can definitely declare tabs and panels within their "inner widget" but I'm not sure whether there's an easy way for us to check their visibility, unless we provide plugin devs with an option for declaring their own context function and I really don't want to go down that path just now.

> My concern here is that it makes things complicated. E.g., when you teach, would you recommend the bind_key decorator but give the caveat that if they want to do more complex things, use register_action? And you would want to clear up potential confusion and make sure users know that both do the same thing, and both will register the action with app-model, register_action just has more functionality. Even with good documentation, there may still be confusion.

I don't feel that this is that complicated to be honest. I think if people "just want to add a keybinding" then they won't care at all that the action is registered with app-model, or what app-model is for that matter. They just want the key sequence to have an effect. If you were teaching general plugin development then you'd want to explain the difference between the two, and mention that you can always use `register_action`, and `bind_key` is there for convenience and grandfathering purposes. I also think even if we didn't reimplement `bind_key`, you'd still have the teaching burden of explaining to people who used to use `bind_key` how to do stuff with `register_action` and what the differences are, so anyway you're back at square one lol. My biggest concern is with documenting the differences. Is there stuff that you could use `bind_key` for before, that now just wouldn't work.

**Comment by lucyleeow** [comment]
Ignore the inner widget thing, I got confused.

I'm fine with the `bind_key`, you know more. I did just notice that you pick context by using `Viewer.bind_key` or `Points.bind_key`, which will just be a bit different via `register_action` (unless we want to keep it the same?)

> My biggest concern is with documenting the differences. Is there stuff that you could use bind_key for before, that now just wouldn't work.

My initial thought was that going forward we may be more restrictive on what is considered a valid keybinding, but I see that Kira already added: [validate_key_binding](https://github.com/kne42/napari/blob/273554b95590cd8abd02c52b7cb47cdad73aaa4c/napari/utils/key_bindings/util.py#L82). What kind of differences were you thinking of?

**Comment by DragaDoncila** [comment]
> What kind of differences were you thinking of?

I wasn't necessarily thinking of any differences we should impose, I was just wondering whether there would be anything we'd like people not to do. But I think I confused myself between the plugin-declared keybindings (which I think we've agreed won't be using `register_action` OR `bind_key`) and the convenience decorators for usage in scripts.
